File: 353.design-snake-game.py
#
# @lc app=leetcode id=353 lang=python3
#
# [353] Design Snake Game
#
from collections import deque
from typing import *

# @lc code=start
class SnakeGame:
    # S: O(n), n := len(food)
    def __init__(self, width: int, height: int, food: List[List[int]]):
        self.w, self.h = width, height
        self.body = deque()
        self.set = set()
        self.score = 0
        self.food = food
        self.food_i = 0  # as food is not a set, use an index to mark visited

        self.body.append((0, 0))
        self.set.add((0, 0))

    # T: O(1)
    def move(self, direction: str) -> int:
        if self.score == -1:
            return -1
        r, c = self.body[0]
        if direction == 'U':
            r -= 1
        elif direction == 'D':
            r += 1
        elif direction == 'L':
            c -= 1
        else:
            c += 1
        new_head = (r, c)

        # Try moving
        pre_tail = self.body.pop()
        self.set.remove(pre_tail)

        if r < 0 or r >= self.h or c < 0 or c >= self.w or new_head in self.set:
            # out of bound, or snake head collides with body
            self.score = -1
            return -1
        
        self.set.add(new_head)
        self.body.appendleft(new_head)

        if self.food_i < len(self.food) and r == self.food[self.food_i][0] and c == self.food[self.food_i][1]:
            # food eaten, snake length + 1, add the old snake tail back 
            self.set.add(pre_tail)
            self.body.append(pre_tail)
            self.food_i += 1
            self.score += 1
            return self.score

        # normal move (no food eaten)
        return self.score

# Your SnakeGame object will be instantiated and called as such:
# obj = SnakeGame(width, height, food)
# param_1 = obj.move(direction)
# @lc code=end

File: 353.design-snake-game.java
/*
 * @lc app=leetcode id=353 lang=java
 *
 * [353] Design Snake Game
 */
import java.util.*;

// @lc code=start
class SnakeGame {
    private int w, h, foodIdx, score;
    private int[][] food;
    Deque<Integer> body;
    Set<Integer> bodySet;  // For fast lookup. (r, c) is converted to one integer r * w + c

    public SnakeGame(int width, int height, int[][] food) {
        w = width;
        h = height;
        this.food = food;
        score = 0;
        body = new ArrayDeque<>();
        bodySet = new HashSet<>();

        // Initially at (0, 0)
        body.offer(0);
        bodySet.add(0);
    }
    
    // T: O(1)
    public int move(String direction) {
        if (score == -1) return -1;  // already game over

        int r = body.peekFirst() / w, c = body.peekFirst() % w;
        switch (direction) {
            case "U": --r; break;
            case "D": ++r; break;
            case "L": --c; break;
            case "R": ++c; break;
        }
        // check if out of bound
        if (r < 0 || r >= h || c < 0 || c >= w) return score = -1;

        // try moving forward
        int newHead = r * w + c;
        bodySet.remove(body.peekLast());
        if (bodySet.contains(newHead)) {
            // collision with itself
            return score = -1;
        }
        // now new head is valid
        bodySet.add(newHead);
        body.offerFirst(newHead);

        // check if eating the current food displayed
        if (foodIdx < food.length && r == food[foodIdx][0] && c == food[foodIdx][1]) {
            bodySet.add(body.peekLast());  // food eaten, old tail doesn't change, add it back to bodySet
            ++foodIdx;
            return ++score;
        }

        // no food eaten, remove previous tail from deque as well
        body.pollLast();
        return score;
    }
}

/**
 * Your SnakeGame object will be instantiated and called as such:
 * SnakeGame obj = new SnakeGame(width, height, food);
 * int param_1 = obj.move(direction);
 */
// @lc code=end

